### PR TITLE
ReviewModel rafactoring. Replaced River with sk-learn

### PR DESCRIPTION
**Pull request by alexandr-z**
- replaced River with sk-learn
- batch processing of docs
- used process_texts_into_docs for SpaCy loading and lang detection
- additional sentence filtering
- other improvements
- renamed files
- CACHE_DEFAULT_TIMEOUT=0 (never expire)

### CODE REVIEW

File: colandr/extensions.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ class _BaseModel(sqlalchemy.orm.DeclarativeBase):
 review_model_cache = flask_caching.Cache(
     config={
         "CACHE_TYPE": "SimpleCache",
-        "CACHE_DEFAULT_TIMEOUT": 3600 * 24,  # 24 hours
+        "CACHE_DEFAULT_TIMEOUT": 0,
```
Comment (bdewilde) [inline colandr/extensions.py:27]: Why change this? 24 hours seems like a long time for a cache TTL!

File: colandr/lib/extractors/review_model.py [hunk 2]
```diff
@@ -0,0 +1,416 @@
+"""Metadata extraction from document full text."""
```
Comment (bdewilde) [inline colandr/lib/extractors/review_model.py:1]: just checking, this is actually a renamed `review_metadata.py`, with some minor changes on top?

File: colandr/lib/extractors/review_model.py [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+            ],
+            remainder='drop'
+        )
+        sgd_classifier = SGDClassifier(
+            loss='log_loss', random_state=42, early_stopping=True,
+            n_iter_no_change=10, alpha=5e-4
+        )
+        self.pipeline = Pipeline([
+            ('preprocessor', preprocessor),
+            ('classifier', MultiOutputClassifier(sgd_classifier))
```
Comment (bdewilde) [inline colandr/lib/extractors/review_model.py:114]: okay, neat! i've never used a multioutput classifier before. this is more or less doing what you were before, with one classifier per field -- right?

File: tests/lib/extractors/test_review_model.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+import pandas as pd
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import MultiLabelBinarizer
+
+from lib.extractors.review_model import (
+    ReviewModel, TrainingData, SingleValue
+)
+
+
+class TestReviewModel(unittest.TestCase):
```
Comment (bdewilde) [inline tests/lib/extractors/test_review_model.py]: we use `pytest` for unit testing, rather than `unittest`. would it be possible to refactor this functionality to follow `pytest` patterns?

File: tests/lib/extractors/test_review_model.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+        mock_tokens = []
+        for word in text.split():
+            token = MagicMock()
+            # A simple rule for the test: if the word is 'is', it's a verb
+            token.pos_ = "VERB" if has_verb and word.lower() == "is" else "NOUN"
+            mock_tokens.append(token)
+
+        mock_sent.__iter__ = MagicMock(return_value=iter(mock_tokens))
+        return mock_sent
```
Comment (bdewilde) [inline tests/lib/extractors/test_review_model.py:30]: Would it not be possible to just parse the example sentences using a `spacy` language pipeline, rather than mocking?

File: tests/lib/extractors/test_review_model.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            "This is a perfectly valid and long sentence for model."
+        )
+        short_sent = self._create_mock_sentence("This sentence is too short.")
+        no_verb_sent = self._create_mock_sentence(
+            "A long title fragment with no actual verb.",
+            has_verb=False
+        )
+        keyword_sent = self._create_mock_sentence(
+            "A long sentence that contains a forbidden keyword like DEBUG."
+        )
```
Comment (bdewilde) [inline tests/lib/extractors/test_review_model.py:46]: wait, why is DEBUG forbidden?

File: tests/lib/extractors/test_review_model.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
+"""Tests for the text metadata extractors."""
```
Comment (bdewilde) [inline tests/lib/extractors/test_review_model.py:1]: It's hard to tell what's actually being tested, since there's so much mocking/patching. Is it possible to create more "complete" test cases, and test functionality with less mocking?

File: colandr/lib/extractors/review_model.py [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+        Args:
+            sent: A spaCy Span object representing a sentence.
+
+        Returns:
+            True if the sentence is valid, False otherwise.
+        """
+        if sent is None:
+            return False
+        sent_text = sent.text.strip()
+        exclude_keywords = {"org.apache", "WARN", "DEBUG"}
```
Comment (bdewilde) [inline colandr/lib/extractors/review_model.py:332]: what's going on here? why are we filtering these words/fragments out?

**Review by bdewilde — COMMENTED** [review]
@alexandr-z I have a few "why?" questions, and one non-trivial ask in the unit tests. Otherwise, I think I follow what you're doing with sklearn -- seems fine to me!

File: colandr/extensions.py (hunk 1 above)
Comment (alexandr-z) [reply colandr/extensions.py:27]: In the original Scala app LRU cache is used without TTL. Initially I added TTL but as model training with real data is time consuming process I decided to return back LRU cache imitation to avoid model retraining too often.
I don't know app use cases, so if you think that TTL should be adjusted please let me know.

File: colandr/lib/extractors/review_model.py (hunk 2 above)
Comment (alexandr-z) [reply colandr/lib/extractors/review_model.py:1]: Yes. In the VCS it is renamed with keeping all history, but github displays it as delete/add files

File: colandr/lib/extractors/review_model.py (hunk 3 above)
Comment (alexandr-z) [reply colandr/lib/extractors/review_model.py:114]: Yes. Since we have multiple labels in Review, this is suitable for our task. It works by fitting one classifier per target variable, allowing for the prediction of multiple outputs simultaneously.

File: tests/lib/extractors/test_review_model.py (hunk 5 above)
Comment (alexandr-z) [reply tests/lib/extractors/test_review_model.py:30]: This will slow down the tests because loading spaCy and parsing are slow.

File: colandr/lib/extractors/review_model.py (hunk 8 above)
Comment (alexandr-z) [reply colandr/lib/extractors/review_model.py:332]: This is from the original Scala implementation https://github.com/datakind/permanent-colandr-back/blob/develop/pdfestrian/src/main/scala/org/datakind/ci/pdfestrian/extraction/review/ReviewModel.scala#L78
Initially I skipped this filtering but when I got prod data I encountered that there are garbage logs in the fulltexts that were generated by the initial PDF parser. So, to avoid model to learn incorrect patterns I added this filtering.

File: tests/lib/extractors/test_review_model.py (hunk 6 above)
Comment (alexandr-z) [reply tests/lib/extractors/test_review_model.py:46]: Because in the prod data we have PDF parser logs that should be filtered

File: tests/lib/extractors/test_review_model.py (hunk 4 above)
Comment (alexandr-z) [reply tests/lib/extractors/test_review_model.py]: Done

File: colandr/extensions.py (hunk 1 above)
Comment (bdewilde) [reply colandr/extensions.py:27]: Okay! This is probably fine for now

File: tests/lib/extractors/test_review_model.py (hunk 5 above)
Comment (bdewilde) [reply tests/lib/extractors/test_review_model.py:30]: Something to work on for later, then!

**Review by bdewilde — APPROVED** [review]
:shipit:
